single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,749 @@
+#!/usr/bin/python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Operator for MongoDB Related Charms."""
+
+from __future__ import annotations
+
+import logging
+from collections.abc import Iterable
+from pathlib import Path
+from typing import TYPE_CHECKING, final
+
+from ops.charm import RelationDepartedEvent
+from ops.framework import Object
+from ops.model import Container, Unit
+from pymongo.errors import PyMongoError, ServerSelectionTimeoutError
+from tenacity import Retrying, stop_after_attempt, wait_fixed
+from typing_extensions import override
+
+from single_kernel_mongo.config.literals import (
+    CONTAINER,
+    MAX_PASSWORD_LENGTH,
+    CharmRole,
+    MongoPorts,
+    Scope,
+    Substrates,
+)
+from single_kernel_mongo.config.logrotate_config import LogRotateConfig
+from single_kernel_mongo.config.relations import RelationNames
+from single_kernel_mongo.config.roles import K8S_MONGO, VM_MONGO
+from single_kernel_mongo.core.operator import OperatorProtocol
+from single_kernel_mongo.core.secrets import generate_secret_label
+from single_kernel_mongo.core.structured_config import MongoDBRoles
+from single_kernel_mongo.events.backups import INVALID_S3_INTEGRATION_STATUS, BackupEventsHandler
+from single_kernel_mongo.events.database import DatabaseEventsHandler
+from single_kernel_mongo.events.password_actions import PasswordActionEvents
+from single_kernel_mongo.events.primary_action import PrimaryActionHandler
+from single_kernel_mongo.events.tls import TLSEventsHandler
+from single_kernel_mongo.exceptions import (
+    ContainerNotReadyError,
+    SetPasswordError,
+    ShardingMigrationError,
+    UpgradeInProgressError,
+    WorkloadExecError,
+    WorkloadNotReadyError,
+    WorkloadServiceError,
+)
+from single_kernel_mongo.managers.backups import BackupManager
+from single_kernel_mongo.managers.config import (
+    CommonConfigManager,
+    LogRotateConfigManager,
+    MongoDBConfigManager,
+    MongoDBExporterConfigManager,
+    MongosConfigManager,
+)
+from single_kernel_mongo.managers.mongo import MongoManager
+from single_kernel_mongo.managers.tls import TLSManager
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.utils.mongo_connection import MongoConnection, NotReadyError
+from single_kernel_mongo.utils.mongodb_users import (
+    BackupUser,
+    MonitorUser,
+    OperatorUser,
+    get_user_from_username,
+)
+from single_kernel_mongo.workload import (
+    get_mongodb_workload_for_substrate,
+    get_mongos_workload_for_substrate,
+)
+from single_kernel_mongo.workload.mongodb_workload import MongoDBWorkload
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm  # pragma: nocover
+
+
+logger = logging.getLogger(__name__)
+
+
+@final
+class MongoDBOperator(OperatorProtocol, Object):
```

Comment:
Related to a previous comment, should this be moved to `AbstractCharm`?

Reply:
As answered to the former comment : No because there will be two operators, one of them being for Mongos (part of this cycle's work) to allow a true "single kernel" library so there is a real meaning for the separation between the charm and the operator :)